can/io/mf4.py

```diff
@@ -0,0 +1,392 @@
+        else:
+            sigs = [(np.array([timestamp]), None), (buffer, None)]
+
+        if msg.is_remote_frame:
+            self._mdf.extend(2, sigs)
+        elif msg.is_error_frame:
+            self._mdf.extend(1, sigs)
+        else:
+            self._mdf.extend(0, sigs)
+
+
+class MF4Reader(BaseIOHandler):
+    """
+    Iterator of CAN messages from a MF4 logging file.
+
+    """
+
+    def __init__(self, file):
+        """
+        :param file: a path-like object or as file-like object to read from
+                        If this is a file-like object, is has to be opened in
+                        binary read mode, not text read mode.
+        """
+        super().__init__(file, mode="rb")
+
+        self._mdf = MDF4(file)
+
+        self.start_timestamp = self._mdf.header.start_time.timestamp()
+
+        masters = [self._mdf.get_master(i, copy_master=False) for i in range(3)]
+
+        masters = [
+            np.core.records.fromarrays((master, np.ones(len(master)) * i))
+            for i, master in enumerate(masters)
+        ]
+
+        self.masters = np.sort(np.concatenate(masters))
+
+    def __iter__(self):
+        standard_counter = 0
+        error_counter = 0
+        rtr_counter = 0
+
+        for timestamp, group_index in self.masters:
+
+            # standard frames
+            if group_index == 0:
+                sample = self._mdf.get(
+                    "CAN_DataFrame",
+                    group=group_index,
+                    raw=True,
+                    record_offset=standard_counter,
+                    record_count=1,
+                )
+
+                if sample["CAN_DataFrame.EDL"] == 0:
+
+                    is_extended_id = bool(sample["CAN_DataFrame.IDE"])
+                    channel = sample["CAN_DataFrame.ID"]
+                    arbitration_id = int(sample["CAN_DataFrame.ID"])
+                    size = int(sample["CAN_DataFrame.DataLength"])
+                    dlc = int(sample["CAN_DataFrame.DLC"])
+                    data = sample["CAN_DataFrame.DataBytes"][0, :size].tobytes()
+
+                    msg = Message(
+                        timestamp=timestamp + self.start_timestamp,
+                        is_error_frame=False,
+                        is_remote_frame=False,
+                        is_fd=False,
+                        is_extended_id=is_extended_id,
+                        channel=channel,
+                        arbitration_id=arbitration_id,
+                        data=data,
+                        dlc=dlc,
+                    )
+
+                else:
+                    is_extended_id = bool(sample["CAN_DataFrame.IDE"])
+                    channel = sample["CAN_DataFrame.ID"]
+                    arbitration_id = int(sample["CAN_DataFrame.ID"])
+                    size = int(sample["CAN_DataFrame.DataLength"])
+                    dlc = FD_DLC2LEN(sample["CAN_DataFrame.DLC"])
```

Comment:
For my own curiosity, how does this syntax work? Isn't `FD_DLC2LEN` a `dict` and not a `callable`?

Reply:
I've fixed this int he new commit